portfolio/src/main/webapp/script.js

```diff
@@ -29,100 +28,201 @@ function addRandomGreeting() {
+    setHidden('loaderid', true);
 }
 
+// clone the document
+var documentCopy = null; 
+
 //*********************** SEARCH BAR FUNCTIONS ***********************
 
 window.onload = function(){
+    documentCopy = document.cloneNode(true);    //storing initial doc
     swapLoaderForContent();
+
+    //Runs search on enter
     document.getElementById('searchbar').onkeydown = function(e){
         if(e.keyCode == 13){
-            runSearch(document.getElementById('searchbar').value.toLocaleLowerCase());
+
+            //i only want single word alphanumerics to go through
+            var retrievedInput = document.getElementById('searchbar').value;
+            const regex = new RegExp(/^[a-z0-9]+$/i);
+            if (regex.test(retrievedInput) === false){
+                console.log("nottoday");
+                shakeSearchBar();
+            } else {
+            runSearch(document.getElementById('searchbar')
+                .value.toLocaleLowerCase());
+            }
         }
     };
 }
 
+/**
+ * Code to shake the search bar
+ */
+function shakeSearchBar(){
+    const el = document.getElementById('searchbar');
+    el.preventDefault;
+    el.classList.remove('searchbarglow')
+    void el.offsetWidth;
+    el.classList.add('searchbarglow');
+}
 
 
 //*********************** SEARCH IMPLEMENTATION ***********************
 
-function retrieveHtml(){
-    var source = document.body.innerHTML;
 
-    if (source != null){
-        return source.toLocaleLowerCase();    //toLocaleLowerCase() to ignore caps
-    } else {
-        alert("source is null");
-    }
-}
+/**
+ * Gets the indices of matches
+ */
+function getIndicesOf(small, big) {
 
+    //no lowercase causes issues with finding the index
+    small = small.toLocaleLowerCase();
+    big = big.toLocaleLowerCase();
 
-//js is pass by value so no need for a copy
-function getClosestTag(index, source){
+    var smallLen = small.length;
+    if (smallLen == 0) {return [];}
 
-    while (source.charAt(index) != '>'){
-        if (index == 0){
-            alert("index hit 0");
-            break;
-        }
-        index--;
+    var startIndex = 0, index, indices = [];
+
+    while ((index = big.indexOf(small, startIndex)) > -1) {
+        indices.push(index);
+        startIndex = index + smallLen;
     }
 
-    var endIndex = index+1; //plus 1 to account for bracket
+    return indices;
+}
 
-    while (source.charAt(index) != '<'){
-        index--;
+/**
+ * Walks the tree and finds matches,
+ * returns index array of matches.
+ * Will optimize to gather straight from tree
+ * rather than push into new arr to save space
+ */
+function treeWalker(node, input){
+    var all = [];
+
+    for (node= node.firstChild; node ; node=node.nextSibling){
+        if (node.nodeType==3){
+            all.push(node);
+        } else {
+            all = all.concat(treeWalker(node, input));
+        }
     }
 
-    var startIndex = index;
+    //find matches
+    var indexArr = [];
```

Comment:
Use `let` and `const` instead of `var`. Fix everywhere.

Reply:
Done. I understand the difference between the 3, but why is it important to change?